##### used conda env giza from ll laptop

## Create and Train an Linear regression Model
### We'll start by creating a simple XGBoost model using Scikit-Learn and train it on diabetes dataset

In [1]:
import numpy as np
import pandas as pd
import xgboost as xgb
# import matplotlib.pyplot as plt
# import seaborn as sns
# import xgboost as xgb

In [2]:
df = pd.read_csv('BETF Final.csv')
df.head(10)

Date  Closing Price   IBIT   FBTC   BITB   ARKB  BTCO  EZBC  BRRR  \
0  11-01-2024    46368.58594  111.7  227.0  237.9   65.3  17.4  50.1  29.4   
1  12-01-2024    42853.16797  386.0  195.3   17.4   39.8  28.4   0.0  20.2   
2  16-01-2024    43154.94531  212.7  102.0   50.2  122.3  31.9   0.0  15.3   
3  17-01-2024    42742.65234  371.4  358.1   68.2   50.3  57.6   1.2   1.2   
4  18-01-2024    41262.05859  145.5  177.9   20.1   41.8  58.8   0.0   9.3   
5  19-01-2024    41618.40625  201.5  222.3   56.7   62.6  63.4   0.0  10.4   
6  22-01-2024    39507.36719  260.6  158.7   41.6   65.0   5.6   4.7   9.7   
7  23-01-2024    39845.55078  160.1  157.7   26.3   61.8   0.0   1.1   0.0   
8  24-01-2024    40077.07422   66.2  125.7   19.1   24.9  19.9   1.2   9.1   
9  25-01-2024    39933.80859  170.7  101.0   20.0   16.1   0.0   0.0   6.5   

   HODL  BTCW   GBTC  Total  
0  10.6   1.0  -95.1  655.3  
1   0.0   0.0 -484.1  203.0  
2   7.3   0.0 -594.4  -52.7  
3   4.8   1.6 -460.6  453.8  
4   2.3   0.0 -582.3 -126.6  
5  14.2   2.9 -590.4   43.6  
6   6.8   0.4 -640.5  -87.4  
7   2.2   0.0 -515.3 -106.1  
8   4.5   0.4 -429.3 -158.3  
9   0.0   0.0 -394.1  -79.8

In [3]:
y = df['Closing Price']

In [4]:
# X = df[['IBIT', 'FBTC', 'BITB', 'ARKB', 'BTCO',
#        'EZBC', 'BRRR', 'HODL', 'BTCW',
#        'GBTC', ]]
X = df[['IBIT']]

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=410)

In [7]:
from sklearn.linear_model import LinearRegression

In [8]:
lm = LinearRegression()
lm.fit(X_train,y_train)

LinearRegression()

In [9]:
predictions = lm.predict(X_test)
predictions

array([59425.33942015, 59394.44762914, 59443.63455851, 59268.88099644,
       58988.55548841, 58877.78492389, 59257.28407813, 59463.42929838,
       59481.52448987, 59191.60153221, 59481.52448987, 59481.52448987,
       58918.7740317 , 59481.52448987, 59237.38936483, 59327.26548171,
       59481.52448987, 59387.84938252, 59375.75259704, 59145.31383241,
       59448.23333646, 59392.5481339 , 59321.46702256, 59370.45400506,
       59481.52448987, 59280.07802101, 59446.03392092, 59289.47552378,
       59461.82972344, 59136.21624995])

# Convert the Model to ONNX Format
### Giza supports ONNX models so you'll need to convert the model to ONNX format. After the model is trained, you can convert it to ONNX format using the skl2onnx library.

#### installng from conda env gave problems seems to work from here itself in Jupyter lab

In [10]:
!pip install skl2onnx

In [11]:
from skl2onnx import convert_sklearn

In [12]:
from skl2onnx.common.data_types import FloatTensorType

In [13]:
# Define the initial types for the ONNX model
initial_type = [('float_input', FloatTensorType([None, X_train.shape[1]]))]

In [14]:
# Convert the scikit-learn model to ONNX
onnx_model = convert_sklearn(lm, initial_types=initial_type)

## Save the model

In [15]:
# Save the ONNX model to a file - will create file - "linear_regression.onnx"
with open("linear_regression2.onnx", "wb") as f:
    f.write(onnx_model.SerializeToString())

In [ ]:
# from giza.zkcook import serialize_model
# serialize_model(lm, "etf_lin1.json")

## Transpile your model to Orion Cairo
### We will use Giza-CLI to transpile our saved model to Orion Cairo.

$ giza transpile linear_regression.onnx --output-path verifiable_lr
[giza][2024-06-04 22:45:09.893] No model id provided, checking if model exists ✅
[giza][2024-06-04 22:45:09.906] Model name is: linear_regression
[giza][2024-06-04 22:45:10.610] Model Created with id -> 702! ✅
[giza][2024-06-04 22:45:11.331] Version Created with id -> 1! ✅
[giza][2024-06-04 22:45:11.344] Sending model for transpilation ✅
[giza][2024-06-04 22:45:44.311] Transpilation is fully compatible. Version compiled and Sierra is saved at Giza ✅
[giza][2024-06-04 22:45:45.120] Downloading model ✅
[giza][2024-06-04 22:45:45.148] model saved at: verifiable_lr




$ giza transpile linear_regression2.onnx --output-path verifiable_lr2
[giza][2024-06-05 00:25:02.804] No model id provided, checking if model exists ✅
[giza][2024-06-05 00:25:02.817] Model name is: linear_regression2
[giza][2024-06-05 00:25:03.650] Model Created with id -> 703! ✅
[giza][2024-06-05 00:25:04.245] Version Created with id -> 1! ✅
[giza][2024-06-05 00:25:04.257] Sending model for transpilation ✅
[giza][2024-06-05 00:25:46.040] Transpilation is fully compatible. Version compiled and Sierra is saved at Giza ✅
[giza][2024-06-05 00:25:47.065] Downloading model ✅
[giza][2024-06-05 00:25:47.106] model saved at: verifiable_lr2

## Deploy an inference endpoint
### Now that our model is transpiled to Cairo we can deploy an endpoint to run verifiable inferences. We will use Giza CLI again to run and deploy an endpoint. Ensure to replace model-id and version-id with your ids provided during transpilation.


$ giza endpoints deploy --model-id 702 --version-id 1
▰▱▱▱▱▱▱ Creating endpoint!
[giza][2024-06-04 22:47:24.658] Endpoint is successful ✅
[giza][2024-06-04 22:47:24.670] Endpoint created with id -> 271 ✅
[giza][2024-06-04 22:47:24.672] Endpoint created with endpoint URL: https://endpoint-giza1-702-1-c662a00e-7i3yxzspbq-ew.a.run.app 🎉

In [ ]:
$ giza endpoints deploy --model-id 703 --version-id 1
▰▰▰▰▰▰▰ Creating endpoint!
[giza][2024-06-05 00:26:55.605] Endpoint is successful ✅
[giza][2024-06-05 00:26:55.615] Endpoint created with id -> 273 ✅
[giza][2024-06-05 00:26:55.617] Endpoint created with endpoint URL: https://endpoint-giza1-703-1-c901199c-7i3yxzspbq-ew.a.run.app 🎉

In [ ]:
$ giza endpoints deploy --size XL --model-id 703 --version-id 1
▰▱▱▱▱▱▱ Creating endpoint!
[giza][2024-06-05 00:31:26.543] Endpoint is successful ✅
[giza][2024-06-05 00:31:26.554] Endpoint created with id -> 274 ✅
[giza][2024-06-05 00:31:26.557] Endpoint created with endpoint URL: https://endpoint-giza1-703-1-1dfb2243-7i3yxzspbq-ew.a.run.app 🎉

### Create and Agent using the CLI # 1 -- size issue

$ giza agents create --model-id 690 --version-id 1 --name etfXGB --description etfXGB
[giza][2024-06-03 19:11:50.498] Creating agent ✅
[giza][2024-06-03 19:11:50.506] Using model id and version id to create agent
[giza][2024-06-03 19:11:50.780] Select an existing account to create the agent.
[giza][2024-06-03 19:11:50.787] Available accounts are:
┌──────────┐
│ Accounts │
├──────────┤
│  giza1   │
└──────────┘
Enter the account name: giza1
{
  "id": 34,
  "name": "etfXGB",
  "description": "etfXGB",
  "parameters": {
    "model_id": 690,
    "version_id": 1,
    "endpoint_id": 261,
    "account": "giza1"
  },
  "created_date": "2024-06-04T02:11:57.598948",
  "last_update": "2024-06-04T02:11:57.598948"
}

### Create and Agent using the CLI # 2 
create again after size is M

$ giza agents create --model-id 690 --version-id 1 --name etfXGB --description etfXGB
[giza][2024-06-04 19:06:18.403] Creating agent ✅
[giza][2024-06-04 19:06:18.411] Using model id and version id to create agent
[giza][2024-06-04 19:06:18.672] Select an existing account to create the agent.
[giza][2024-06-04 19:06:18.676] Available accounts are:
┌──────────┐
│ Accounts │
├──────────┤
│  giza1   │
└──────────┘
Enter the account name: giza1
{
  "id": 37,
  "name": "etfXGB",
  "description": "etfXGB",
  "parameters": {
    "model_id": 690,
    "version_id": 1,
    "endpoint_id": 269,
    "account": "giza1"
  },
  "created_date": "2024-06-05T02:06:25.417540",
  "last_update": "2024-06-05T02:06:25.417540"
}

### Create and Agent using the CLI # 2 
create again after size is XL

$ giza agents create --model-id 690 --version-id 1 --endpoint-id 270 --name etfXGB2 --description etfXGB2
[giza][2024-06-04 19:24:46.361] Creating agent ✅
[giza][2024-06-04 19:24:46.370] Using endpoint id to create agent, retrieving model id and version id
[giza][2024-06-04 19:24:46.626] Select an existing account to create the agent.
[giza][2024-06-04 19:24:46.629] Available accounts are:
┌──────────┐
│ Accounts │
├──────────┤
│  giza1   │
└──────────┘
Enter the account name: giza1
{
  "id": 40,
  "name": "etfXGB2",
  "description": "etfXGB2",
  "parameters": {
    "model_id": 690,
    "version_id": 1,
    "endpoint_id": 270,
    "account": "giza1"
  },
  "created_date": "2024-06-05T02:24:51.304557",
  "last_update": "2024-06-05T02:24:51.304557"
}

$ giza agents create --model-id 703 --version-id 1 --endpoint-id 274 --name etfLIN1 --description etfLIN1
[giza][2024-06-05 00:48:43.649] Creating agent ✅
[giza][2024-06-05 00:48:43.669] Using endpoint id to create agent, retrieving model id and version id
[giza][2024-06-05 00:48:43.948] Select an existing account to create the agent.
[giza][2024-06-05 00:48:43.957] Available accounts are:
┌──────────┐
│ Accounts │
├──────────┤
│  giza1   │
└──────────┘
Enter the account name: giza1
{
  "id": 41,
  "name": "etfLIN1",
  "description": "etfLIN1",
  "parameters": {
    "model_id": 703,
    "version_id": 1,
    "endpoint_id": 274,
    "account": "giza1"
  },
  "created_date": "2024-06-05T07:48:47.721178",
  "last_update": "2024-06-05T07:48:47.721178"
}

In [ ]:
X_test

In [16]:
# new_data = pd.DataFrame({
#     'IBIT': [120],
#     'FBTC': [175],
#     'BITB': [5],
#     'ARKB': [50],
#     'BTCO': [1],
#     'EZBC': [2],
#     'BRRR': [3],
#     'HODL': [4],
#     'BTCW': [5],
#     'GBTC': [-20]
# })

new_data = pd.DataFrame({
    'IBIT': [120]
})

In [17]:
# new_data = pd.DataFrame({
#     'IBIT': [120],
#     'FBTC': [175],
#     'BITB': [5],
#     'ARKB': [50],
#     'BTCO': [1],
#     'EZBC': [2],
#     'BRRR': [3],
#     'HODL': [4],
#     'BTCW': [5],
#     'GBTC': [-20]
# })

# Predicting the expenses for the new data
predicted_btc_closing_price = lm.predict(new_data)

print(predicted_btc_closing_price )

[59361.55636946]


## Run a verifiable inference
##### To streamline verifiable inference, you might consider using the endpoint URL obtained after transpilation. However, this approach requires manual serialization of the input for the Cairo program and handling the deserialization process. To make this process more user-friendly and keep you within a Python environment, we've introduced a Python SDK designed to facilitate the creation of ML workflows and execution of verifiable predictions. When you initiate a prediction, our system automatically retrieves the endpoint URL you deployed earlier, converts your input into Cairo-compatible format, executes the prediction, and then converts the output back into a numpy object. 

In [21]:
from giza.agents.model import GizaModel

MODEL_ID = 703  # Update with your model ID
VERSION_ID = 1  # Update with your version ID

def prediction(input, model_id, version_id):
    model = GizaModel(id=model_id, version=version_id)

    (result, proof_id) = model.predict(
        input_feed={'input': input}, verifiable=True
    )

    return result, proof_id

def execution():
    # The input data type should match the model's expected input
    # input = np.array([[5.5]]).astype(np.float32)
    new_data = pd.DataFrame({
    'IBIT': [120]
})


    # (result, proof_id) = prediction(input, MODEL_ID, VERSION_ID)
    (result, proof_id) = prediction(new_data, MODEL_ID, VERSION_ID)

    print(
        # f"Predicted value for input {input.flatten()[0]} is {result[0].flatten()[0]}")
        f"Predicted value for input is {result[0].flatten()[0]}")

    return result, proof_id


execution()

An error occurred in predict: 500 Server Error: Internal Server Error for url: https://endpoint-giza1-703-1-1dfb2243-7i3yxzspbq-ew.a.run.app/cairo_run
Deployment predict error: Running the Cairo  programm failed
An error occurred in predict: 500 Server Error: Internal Server Error for url: https://endpoint-giza1-703-1-1dfb2243-7i3yxzspbq-ew.a.run.app/cairo_run


HTTPError: 500 Server Error: Internal Server Error for url: https://endpoint-giza1-703-1-1dfb2243-7i3yxzspbq-ew.a.run.app/cairo_run

## Download the proof
#### Initiating a verifiable inference sets off a proving job on our server, sparing you the complexities of installing and configuring the prover yourself. Upon completion, you can download your proof.

First, let's check the status of the proving job to ensure that it has been completed.

$ giza endpoints get-proof --endpoint-id 234 --proof-id "10c164e6c2364ab6b5491702127979a6"
[giza][2024-05-30 00:40:39.691] Getting proof from endpoint 234 ✅
{
  "id": 967,
  "job_id": 1121,
  "metrics": {
    "proving_time": 17.249508
  },
  "created_date": "2024-05-30T07:33:12.532659"
}

Once the proof is ready, you can download it.

$ giza endpoints download-proof --endpoint-id 234 --proof-id "10c164e6c2364ab6b5491702127979a6" --output-path zk_xgboost.proof
[giza][2024-05-30 00:51:52.048] Getting proof from endpoint 234 ✅
[giza][2024-05-30 00:51:53.800] Proof downloaded to zk_xgboost.proof ✅
(giza3)

## Verify the proof
#### Finally, you can verify the proof.

$ giza verify --proof-id 967
[giza][2024-05-30 00:56:05.847] Verifying proof...
[giza][2024-05-30 00:56:07.140] Verification result: True
[giza][2024-05-30 00:56:07.145] Verification time: 0.454667226
(giza3)